teku/src/main/java/tech/pegasys/teku/cli/util/DatabaseMigrater.java

```diff
@@ -0,0 +1,256 @@
+    this.spec = spec;
+    this.batchSize = batchSize;
+    this.statusUpdater = statusUpdater;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public void migrateDatabase(
+      final DatabaseVersion sourceDatabaseVersion, final DatabaseVersion targetDatabaseVersion)
+      throws DatabaseMigraterError {
+    if (migratedDatabasePath.isEmpty()) {
+      try {
+        duplicateBeaconFolderContents();
+      } catch (IOException ex) {
+        throw new DatabaseMigraterError(
+            "Failed to create new database structure: " + ex.getMessage());
+      }
+    }
+
+    openDatabases(sourceDatabaseVersion, targetDatabaseVersion);
+
+    statusUpdater.accept("Migrating data to the new database");
+    migrateData();
+
+    closeDatabases();
+    statusUpdater.accept("Swapping new database to be active");
+    swapActiveDatabase();
+  }
+
+  @VisibleForTesting
+  void openDatabases(
+      final DatabaseVersion sourceDatabaseVersion, final DatabaseVersion targetDatabaseVersion) {
+    final Path newDatabasePath = getNewBeaconFolderPath();
+    final Path originalDatabasePath = dataDirLayout.getBeaconDataDirectory();
+    statusUpdater.accept("Opening original database...");
+    originalDatabase = createDatabase(originalDatabasePath, sourceDatabaseVersion);
+    statusUpdater.accept("Creating a new database...");
+    newDatabase = createDatabase(newDatabasePath, targetDatabaseVersion);
+  }
+
+  @VisibleForTesting
+  void swapActiveDatabase() throws DatabaseMigraterError {
+    try {
+      Files.move(
+          dataDirLayout.getBeaconDataDirectory(),
+          getMovedOldBeaconFolderPath(),
+          StandardCopyOption.ATOMIC_MOVE);
+    } catch (IOException ex) {
+      statusUpdater.accept(ex.getMessage());
+      throw new DatabaseMigraterError(
+          "Failed to move old database to " + getMovedOldBeaconFolderPath().toString());
+    }
+    try {
+      Files.move(
+          getNewBeaconFolderPath(),
+          dataDirLayout.getBeaconDataDirectory(),
+          StandardCopyOption.ATOMIC_MOVE);
+    } catch (IOException ex) {
+      statusUpdater.accept(ex.getMessage());
+      throw new DatabaseMigraterError(
+          "Failed to move new database to " + dataDirLayout.getBeaconDataDirectory().toString());
+    }
+  }
+
+  @VisibleForTesting
+  void closeDatabases() {
+    try {
+      originalDatabase.close();
+    } catch (Exception e) {
+      statusUpdater.accept("Failed to close original database cleanly: " + e.getMessage());
+    }
+    try {
+      newDatabase.close();
+    } catch (Exception e) {
+      statusUpdater.accept("Failed to close new database cleanly: " + e.getMessage());
+    }
+  }
+
+  @VisibleForTesting
+  void duplicateBeaconFolderContents() throws IOException {
+    final Path newBeaconFolderPath = getNewBeaconFolderPath();
+    if (newBeaconFolderPath.toFile().isDirectory()) {
+      FileUtils.deleteDirectory(newBeaconFolderPath.toFile());
+    }
```

Comment:
Didn't we abort back up in the subcommand if the `.new` directory already existed? So we shouldn't need this right?

Reply:
if the 'new' folder existed, we just swallow it...
use case is when my buffer sizes were too big, i didn't have to delete the .new each time i ran, just tweak the batch-size...

if the 'old' folder exists, it means we've done a migrate fully, and I'm not so keen on deleting what should have been a successful conversion...

Reply:
added some comments around how folders are used to hopefully explain the reasoning behind which folders we care about and which we don't.